LANGUAGE_SDK_SPECIFICATION.md

```diff
@@ -0,0 +1,1325 @@
+
+#### 4.8.2 Required Behavior
+
+The SDK MUST:
+
+1. Support nesting operations within a context
+2. Associate child operations with the context via `ParentId`
+3. Propagate context completion to all children
+4. Support conditional replay of children
+5. Enable context-level error handling
+
+#### 4.8.3 Supported Actions
+
+- `START`: Begin a new context
+- `SUCCEED`: Complete the context successfully
+- `FAIL`: Complete the context with an error
+
+#### 4.8.4 State Transitions
+
+```
+[New] → START → STARTED → SUCCEED → SUCCEEDED [Done]
+                        ↓
+                        └→ FAIL → FAILED [Done]
+```
+
+#### 4.8.5 Replay Optimization
+
+The SDK SHOULD support `ReplayChildren` option:
+
+- When `false` and context is completed, child operations are not included in future state loads
+- When `true`, child operations are retained for replay
+
+This optimization is RECOMMENDED for large parallel operations and maps.
+
+## 5. Checkpointing
+
+### 5.1 Checkpoint API
+
+The SDK MUST use the `CheckpointDurableExecution` API to:
+
+- Start new operations
+- Complete existing operations (succeed, fail, retry)
+- Update operation state
+
+### 5.2 Checkpoint Tokens
+
+The SDK MUST:
+
+1. Use the `CheckpointToken` from the invocation input for the first checkpoint
+2. Use the returned `CheckpointToken` from each checkpoint response for subsequent checkpoints
+3. Handle `InvalidParameterValueException` errors for invalid tokens
+4. Never reuse a checkpoint token that has already been consumed
+
+### 5.3 Checkpoint Request Format
+
+Each checkpoint request MUST include:
+
+- `DurableExecutionArn`: The execution identifier
+- `CheckpointToken`: The current valid token
+- `Updates`: Array of operation updates
+
+### 5.4 Operation Updates
+
+Each operation update MUST include:
+
+- `Id`: Unique identifier for the operation
+- `Type`: Operation type
+- `Action`: Action to perform
+
+Each operation update MAY include:
+
+- `Name`: Human-readable name
+- `ParentId`: Parent context ID
+- `SubType`: SDK-level categorization
+- `Payload`: Success result (JSON string)
+- `Error`: Failure error object
+- Type-specific options (`StepOptions`, `WaitOptions`, `CallbackOptions`, `ChainedInvokeOptions`, `ContextOptions`)
+
+For detailed field constraints (length limits, patterns, and valid values), refer to the AWS Lambda API documentation:  
+https://docs.aws.amazon.com/lambda/latest/api/API_OperationUpdate.html
+
+### 5.5 Checkpoint Response
+
+The SDK MUST process checkpoint responses that include:
+
+- `CheckpointToken`: New token for next checkpoint
+- `NewExecutionState.Operations`: Updated operation state
+- `NewExecutionState.NextMarker`: Pagination token (if applicable)
+
+### 5.6 Checkpoint Frequency
+
+The SDK SHOULD checkpoint:
+
+- After each durable operation starts
+- After each durable operation completes
+- Before returning with pending status
```

Comment:
It doesn't need to checkpoint before returning PENDING status if it already knows there are only pending operations remaining